pypowerwall/scan.py

```diff
@@ -8,200 +8,294 @@
+    """Thread Worker: Scan IP Address for presence of a Tesla Energy Gateway.
+
+    Args:
+        addr (IPv4Address): IP address to scan
+        context (ScanContext): Context controlling output interactivity, color, and timeout behaviors.
+        result_queue (Queue): Thread safe queue to store results of the asynchronous scan.
     """
-    pyPowerwall Network Scanner
 
-    Parameter:
-        color = True or False, print output in color [Default: True]
-        timeout = Seconds to wait per host [Default: 1.0]
-        hosts = Number of hosts to scan simultaneously [Default: 30]
-        ip = IP address within network to scan [Default: None (detect IP)]
+    host = f"{context.dim()}\r\t  Host: {context.subbold()}{addr} ...{context.normal()}"
+    if context.interactive:
+        print(host, end='')
 
-    Description
-            This tool will scan your local network looking for a Tesla Energy Gateway
-            and Powerwall.  It tries to use your local IP address as a default.
+    if not check_connection(addr, context):
+        return
 
+    # noinspection PyBroadException
+    try:
+        # Check to see if it is a Powerwall
+        response = requests.get(f'https://{addr}/api/status', verify=False, timeout=5)
+        if response.status_code != HTTPStatus.NOT_FOUND:
+            data: Final[str] = response.json()
+            din: Final[str] = data.get('din', UNKNOWN)
+            version: Final[str] = data.get('version', UNKNOWN)
+            if din == UNKNOWN and version == UNKNOWN:
+                return
+
+            up_time: Final[str] = data.get('up_time_seconds', UNKNOWN)
+            type_selector = lambda type: type if not caseless_equal(type, UNKNOWN) else POWERWALL
+            teg_type: Final[str] = type_selector(data.get('teg_type', POWERWALL))
+
+            if context.interactive:
+                print(f"{host} OPEN{context.dim()} - {context.subbold()}Found {teg_type} {din}{context.subbold()}\n\t\t\t\t\t\t\t\t\t [Firmware {version}]{context.normal()}")
+            result_queue.put({
+                'ip': addr,
+                'din': din,
+                'firmware': version,
+                'up_time': up_time
+            })
+            return
+
+        # Check if it is a Powerwall 3
+        response_pw3 = requests.get(f'https://{addr}/tedapi/din', verify=False, timeout=5)
+        # Expected response from PW3
+        # {"code":403,"error":"Unable to GET to resource","message":"User does not have adequate access rights"}
+        if "User does not have adequate access rights" in response_pw3.text:
+            # Found PW3
+            if context.interactive:
+                print(f"{host} OPEN{context.dim()} - {context.subbold()}Found Powerwall 3 [Supported in Cloud Mode only]{context.normal}")
+            result_queue.put({
+                'ip': addr,
+                'din': 'Powerwall-3',
+                'firmware': 'Supported in Cloud Mode only - See https://tinyurl.com/pw3support'
```

Comment:
Since we are updating, I'll update this to include insturctions that PW3 users can use cloud or TEDAPI mode.